protocol/grpc_new/client.go

```diff
@@ -0,0 +1,167 @@
+type Client struct {
+	*http.Client
+	invoker reflect.Value
+}
+
+// NewClient creates a new GRPC_NEW client.
+func NewClient(url *common.URL) (*Client, error) {
+	// If global trace instance was set, it means trace function enabled.
+	// If not, will return NoopTracer.
+	//tracer := opentracing.GlobalTracer()
+	var cliOpts []connect.ClientOption
+	// set max send and recv msg size
+	maxCallRecvMsgSize := constant.DefaultMaxCallRecvMsgSize
+	if recvMsgSize, err := humanize.ParseBytes(url.GetParam(constant.MaxCallRecvMsgSize, "")); err == nil && recvMsgSize > 0 {
+		maxCallRecvMsgSize = int(recvMsgSize)
+	}
+	cliOpts = append(cliOpts, connect.WithReadMaxBytes(maxCallRecvMsgSize))
+	maxCallSendMsgSize := constant.DefaultMaxCallSendMsgSize
+	if sendMsgSize, err := humanize.ParseBytes(url.GetParam(constant.MaxCallSendMsgSize, "")); err == nil && sendMsgSize > 0 {
+		maxCallSendMsgSize = int(sendMsgSize)
+	}
+	cliOpts = append(cliOpts, connect.WithSendMaxBytes(maxCallSendMsgSize))
+
+	// consumer config client connectTimeout
+	//connectTimeout := config.GetConsumerConfig().ConnectTimeout
+
+	//dialOpts = append(dialOpts,
+	//	grpc.WithBlock(),
+	//	// todo config network timeout
+	//	grpc.WithTimeout(time.Second*3),
+	//	grpc.WithUnaryInterceptor(otgrpc.OpenTracingClientInterceptor(tracer, otgrpc.LogPayloads())),
+	//	grpc.WithStreamInterceptor(otgrpc.OpenTracingStreamClientInterceptor(tracer, otgrpc.LogPayloads())),
+	//	grpc.WithDefaultCallOptions(
+	//		grpc.CallContentSubtype(clientConf.ContentSubType),
+	//		grpc.MaxCallRecvMsgSize(maxCallRecvMsgSize),
+	//		grpc.MaxCallSendMsgSize(maxCallSendMsgSize),
+	//	),
+	//)
+	var cfg *tls.Config
+	var tlsFlag bool
+	var err error
+	tlsConfig := config.GetRootConfig().TLSConfig
+	if tlsConfig != nil {
+		cfg, err = config.GetClientTlsConfig(&config.TLSConfig{
+			CACertFile:    tlsConfig.CACertFile,
+			TLSCertFile:   tlsConfig.TLSCertFile,
+			TLSKeyFile:    tlsConfig.TLSKeyFile,
+			TLSServerName: tlsConfig.TLSServerName,
+		})
+		logger.Infof("GRPC_NEW Client initialized the TLSConfig configuration")
+		if err != nil {
+			return nil, err
+		}
+		tlsFlag = true
+	}
+
+	key := url.GetParam(constant.InterfaceKey, "")
+	impl := config.GetConsumerServiceByInterfaceName(key)
+	conRefs := config.GetConsumerConfig().References
+	ref, ok := conRefs[key]
+	if !ok {
+		panic("no reference")
+	}
+	// todo: set timeout
+	var transport http.RoundTripper
+	switch ref.Protocol {
+	case "http1":
+		transport = &http.Transport{
+			TLSClientConfig: cfg,
+		}
+		cliOpts = append(cliOpts, connect.WithHTTPGet())
+	case GRPC_NEW:
+		if tlsFlag {
+			transport = &http2.Transport{
+				TLSClientConfig: cfg,
+			}
+		} else {
+			transport = &http2.Transport{
+				DialTLSContext: func(_ context.Context, network, addr string, _ *tls.Config) (net.Conn, error) {
+					return net.Dial(network, addr)
+				},
+				AllowHTTP: true,
+			}
+		}
+		cliOpts = append(cliOpts, connect.WithGRPC())
+	// todo: process triple
```

Comment:
Will the GPRC_NEW replace the Triple protocol?

Reply:
This depends on how the composition of the triple protocol is identified. Maybe Rest(?)+GRPC ?